Make date range accept start and stop times

tykeal opened the pull request:
## Breaking change
None

## Proposed change
Sometimes you want to have more granular control over the allowed
window when a code is active.

For instance, you want to hand out a single use code, but it can only be
used on a certain day two weeks in the future during a given set of
hours.

Or, you have a guest coming that you want to give them an access code,
but you know that they won't be there before say, 13:00 on a given day
and they'll be leaving by 15:00 on a day in the following week.

In any of these cases you don't want the code to be active all day on
either of the days that they are arriving or leaving, but you also can't
just use a date range and limit the hours using the more granular by day
options.

## Type of change
* [ ]  Dependency upgrade
* [ ]  Bugfix (non-breaking change which fixes an issue)
* [x]  New feature (which adds functionality)
* [ ]  Breaking change (fix/feature causing existing functionality to break)
* [ ]  Code quality improvements to existing code or addition of tests

## Additional information
* This PR fixes or closes issue: fixes #144
* This PR is related to issue:

Signed-off-by: Andrew Grimberg <tykeal@bardicgrove.org>

tykeal commented:
@FutureTense I didn't make any changes to the imports, just the templates. I dunno why it's failing the tests on importing!

firstof9 commented:
They are failing due to a library.

Please use the PR template as well.

tykeal commented:
@firstof9 ok, I've updated the PR to use the template.

firstof9 commented:
You've tested this locally and it works as expected?

tykeal commented:
I tested the code changes against a specific slot on my lock.

So, what I did was test rendered template but only on one slot so I could verify new functionality against present functionality

tykeal commented:
Just an update, I'm going to do a full validation right now by pushing the code to my installation and removing and regenerating my lock configuration

firstof9 commented:
It look good to me, but I'd like a second set of eyes on it.

tykeal commented:
I totally understand!

raman325 commented:
will test the functionality later. My only gripe so far is that now, by default, the range will end on midnight of the last day which means the last day isn't included. I'm not sure if there's a way around that, just wanted to point it out

tykeal commented:
@raman325  umm... from my reading of the previous logic that's exactly what was happening

If start and end were the same day, then it would enable the lock from $day 00:00 to $day+1 00:00

If start and end were on differing days then it would enable the lock from $startday 00:00 to $endday 00:00 (so early morning of the endday)


Now, for some reason, now that I applied the full code set to my instance, removed and re-added my lock the same day logic doesn't seem to be firing properly. So I'm going to debug that.

tykeal commented:
@raman325 ahh... I think I see your issue now. Let me rework the logic so that if the times are the same (not just the timestamps) that we evaluate the range properly. It's just an extra logic step

tykeal committed:
Make date range accept start and stop times

Sometimes you want to have more granular control over the allowed
window when a code is active.

For instance, you want to hand out a single use code, but it can only be
used on a certain day two weeks in the future during a given set of
hours.

Or, you have a guest coming that you want to give them an access code,
but you know that they won't be there before say, 13:00 on a given day
and they'll be leaving by 15:00 on a day in the following week.

In any of these cases you don't want the code to be active all day on
either of the days that they are arriving or leaving, but you also can't
just use a date range and limit the hours using the more granular by day
options.

Fixes: #144
Signed-off-by: Andrew Grimberg <tykeal@bardicgrove.org>

tykeal commented:
@firstof9 and @raman325 I've pushed an updated version. Please note I also added in a comment at the top to disable yamllint as my code editor was just massively unhappy about all the yaml cleanliness issues of this!

I've confirmed that this version appears to be working as intended including the backwards compatibility that @raman325 was asking for, though it's specific to 00:00 on both start and end dates. If the time is any time _other_ than that, it will end up in the standard date time range evaluation. This way, if someone really does want a code to start at midnight and end at midnight, they just need a 1 minute difference on that to make the range work.

raman325 commented:
> @firstof9 and @raman325 I've pushed an updated version. Please note I also added in a comment at the top to disable yamllint as my code editor was just massively unhappy about all the yaml cleanliness issues of this!
> 
> I've confirmed that this version appears to be working as intended including the backwards compatibility that @raman325 was asking for, though it's specific to 00:00 on both start and end dates. If the time is any time _other_ than that, it will end up in the standard date time range evaluation. This way, if someone really does want a code to start at midnight and end at midnight, they just need a 1 minute difference on that to make the range work.

sorry to do this to you but do you still have the older version (the one that didn't do the special handling for the last day if time = 00:00:00)? I think simpler is better, and while it does change the behavior, I think it's much simpler to explain that the time range affects the filter than to explain that the time range affects the filter **except** for when it's exactly midnight then you get an additional day.

tykeal commented on custom_components/keymaster/keymaster.yaml:
```diff
@@ -207,15 +207,28 @@ binary_sensor:
+        {% set end_time = (states('input_datetime.end_date_LOCKNAME_TEMPLATENUM')[11:]).replace(':', '') | int %}
+
+        {## If the start and end dates match, including time, then we do a basic check against just today ##}
+        {% if (start_date == end_date) %}
```
@raman325 is this the check that you're talking about? I can just rip this first one out and make the elif the first if statement, or better (and IMO actually better) just use the final check clause which means that the date and time are relevant always. I personally like this better as it's more what I'm familiar with when it comes to these sort of lock management things.

That last one does mean that the defaults wouldn't work since they always default to same day and 00:00 which is why I actually have this first check in there. For those times when you just want to give someone access for a given day...

raman325 replied:
Yes, I am suggesting the same as you, that we eliminate all of these conditionals and just use the final clause. For the case where a user wants to enable the slot for a single day, I don't think it's a lot to ask them to put in 23:59:59, especially since you have to explicitly enable the date range check for it to be included. This makes the logic simple and consistent in every scenario, which I think will ultimately be better

raman325 replied:
If we do this, I think we can also get rid of `current_datestamp`, `start_time`, and `end_time` and make this massive logic block just a teensy bit smaller 🙂

tykeal commented on custom_components/keymaster/keymaster.yaml:
```diff
@@ -207,15 +207,28 @@ binary_sensor:
+        {% if (start_date == end_date) %}
+        {%   set is_in_date_range = (current_datestamp == start_date) %}
+        {## Make backwards compatible such that if start & end date times are midnight, we do a day based range only ##}
+        {% elif (start_date < end_date and start_time == end_time and start_time == 0) %}
```
My testing of this logic has shown me that it works similar to the original code

raman325 replied:
See above

tykeal commented:
> sorry to do this to you but do you still have the older version (the one that didn't do the special handling for the last day if time = 00:00:00)? I think simpler is better, and while it does change the behavior, I think it's much simpler to explain that the time range affects the filter than to explain that the time range affects the filter **except** for when it's exactly midnight then you get an additional day.

The older version didn't have the elif statement in it. I can revert back to that with a little work. You may have noticed I force pushed my update... that's because I usually use Gerrit and it retains all intermediary to final patches automatically but GitHub doesn't :( Plus I tend to prefer clean merges of a final product and really dislike squash merge since I GPG sign my changes!

raman325 commented:
> The older version didn't have the elif statement in it. I can revert back to that with a little work. You may have noticed I force pushed my update... that's because I usually use Gerrit and it retains all intermediary to final patches automatically but GitHub doesn't :( Plus I tend to prefer clean merges of a final product and really dislike squash merge since I GPG sign my changes!

It's just easier to review individual commits so that we don't have to review the entire change again, and it makes it easier to revert in cases like this, but it's more of a burden on you than it is on me so if that's what you prefer, I will work with it!

raman325 commented:
You may want to rebase on master, that will resolve the issue with tests

tykeal commented:
Ok, latest version is based on most recent main. Also, this version only checks the full datetime range. To get something similar to what used to happen when you first enabled the range and had start and end days equal you need to now do startday and startday+1 which will effectively give the same functionality as before, just requires people to actually do a little more configuration than flicking the use date range toggle.

raman325 reviewed (approved):
thanks for iterating through this with me!

tykeal commented:
@raman325 no worries, I'm "that person" asking for changes in other projects ;)
